modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,134 @@
+package org.tessellation.dag.l0.domain.snapshot.programs
+
+import cats.MonadThrow
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.option._
+import cats.syntax.traverse._
+import cats.syntax.unorderedFoldable._
+
+import org.tessellation.dag.l0.infrastructure.snapshot.{DAGEvent, GlobalSnapshotEvent, StateChannelEvent}
+import org.tessellation.ext.kryo._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.node.shared.domain.event.EventCutter
+import org.tessellation.schema.Block
+import org.tessellation.schema.gossip.Ordinal
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.security.hash._
+import org.tessellation.security.{Hashed, Hasher}
+
+import eu.timepit.refined.types.numeric.PosInt
+
+object GlobalSnapshotEventCutter {
+  type GlobalPeerIdEventMap = Map[PeerId, List[(Ordinal, GlobalSnapshotEvent)]]
+
+  def make[F[_]: Async: KryoSerializer: Hasher](maxBinarySizeBytes: PosInt): EventCutter[F, PeerId, Ordinal, GlobalSnapshotEvent] =
+    new EventCutter[F, PeerId, Ordinal, GlobalSnapshotEvent] {
+
+      def cut(events: GlobalPeerIdEventMap): F[GlobalPeerIdEventMap] =
+        needsCutting(events)
+          .ifM(
+            OrderedEvents(events)
+              .map(Iterator.unfold(_)(_.next))
```

Comment:
I don't understand; that's what the iterator is doing, cutting from a pre-calculated order.

Coming up with a definitive order for removal with a one-time sort might be difficult/confusing due to the no. of events rule. Maybe the re-sorting can be eliminated if we rework the algorithm to ignore parents rather than remove them. (The first pass at this was checking the lengths of the resulting list, not the original with the parents.) However, I wonder if any time-savings is worth the added complication used in removal. 🤔

Reply:
@TheMMaciek I'll add another comment since I haven't changed the selection function:

We might be able to come up with a one-time sort and then pick events from that. However, I think the sorting algorithm maybe complicated because of the criteria of not removing parents and the no. of each event (i.e. length of the `Left` and `Right` lists.)

My understanding is this cutting shouldn't happen that often so my thought was the keep the code easier to follow.

WDYT?